Add graphic pipeline related source code

kexia01 opened the pull request.

kexia01 committed:
Add renderpass setup and some vulkan resources for graphic pipeline

kexia01 committed:
Add graphic pipeline related sources;
Add one basic graphic test case.

kexia01 committed:
Add graphic pipeline related sources;
Add one basic graphic test case.

kexia01 committed:
Merge branch 'main' of https://github.com/kexia01/tracetooltests

per-mathisen-arm commented on CMakeLists.txt:
```diff
@@ -63,12 +63,12 @@ endfunction()
 	target_compile_definitions(vulkan_${ARGV0} PUBLIC ${IT_DEFINES})
 	set_target_properties(vulkan_${ARGV0} PROPERTIES COMPILE_FLAGS ${IT_CFLAGS})
-	target_include_directories(vulkan_${ARGV0} PUBLIC ${PROJECT_SOURCE_DIR}/include ${PROJECT_SOURCE_DIR} ${PROJECT_SOURCE_DIR}/external/Vulkan-Headers/include)
+	target_include_directories(vulkan_${ARGV0} PUBLIC ${PROJECT_SOURCE_DIR}/include ${PROJECT_SOURCE_DIR} ${PROJECT_SOURCE_DIR}/external ${PROJECT_SOURCE_DIR}/external/Vulkan-Headers/include)
```
Why do you need to load includes from the base external dir? if it is for stb_image_write.h then you could just prepend it with "external/".

kexia01 replied:
right.

per-mathisen-arm commented on src/vulkan_graphics_1.cpp:
```diff
@@ -0,0 +1,399 @@
+    // contexts and resources related with benchmark
+    uint32_t width = 0 ;
+    uint32_t height = 0;
+    std::shared_ptr<Buffer> m_vertexBuffer;
```
Do they really need to be shared pointers?

kexia01 replied:
Pointer members of a class may be passed as function parameters multiple times. 
Using a unique pointer will transfer ownership of the object.
For local variables, unique pointers could be used, but developers should handle them carefully.

kexia01 replied:
change to unique pointer in test case as much as possible, while also modifying the parameters type.

per-mathisen-arm commented on src/vulkan_graphics_1.cpp:
```diff
@@ -0,0 +1,399 @@
+	req.cmdopt = test_cmdopt;
+	vulkan_setup_t vulkan = test_init(argc, argv, "vulkan_graphics_1", req);
+
+    p__loops = 1;
```
Please fix indentation issues

per-mathisen-arm commented on src/vulkan_graphics_1.cpp:
```diff
@@ -0,0 +1,399 @@
+    /******************** sampled texture image for background **********************/
+    // loading image
+    int texWidth, texHeight, texChannels;
+    stbi_uc* pixels = stbi_load("texture/girl.jpg", &texWidth, &texHeight, &texChannels, STBI_rgb_alpha);
```
This file seems to be missing in this commit?

per-mathisen-arm commented on src/vulkan_graphics_1.cpp:
```diff
@@ -0,0 +1,399 @@
+    bgImage->create( {(uint32_t)texWidth, (uint32_t)texHeight, 1}, VK_FORMAT_R8G8B8A8_SRGB, 
+        VK_IMAGE_USAGE_TRANSFER_DST_BIT | VK_IMAGE_USAGE_SAMPLED_BIT, VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT);
+    auto bgImageView = std::make_shared<ImageView>(bgImage);
+    bgImageView->create(VK_IMAGE_VIEW_TYPE_2D /*, VK_IMAGE_ASPECT_COLOR_BIT*/);
```
Please no commented out code :-)

per-mathisen-arm commented on src/vulkan_graphics_common.cpp:
```diff
@@ -0,0 +1,1454 @@
+	printf("-i/--image-output      Save an image of the output to disk\n");
+	printf("-W/--width             Width of output image (default 640)\n");
+	printf("-H/--height            Height of output image (default 480)\n");
+	printf("-wg/--workgroup-size   Set workgroup size (default 32)\n");
```
I don't think this is used here?

kexia01 replied:
remove -i,-wg

per-mathisen-arm commented on src/vulkan_graphics_common.cpp:
```diff
@@ -0,0 +1,1454 @@
+
+    m_memoryProperty = properties;
+
+    {
```
Not sure why this brace is here?

per-mathisen-arm commented on src/vulkan_graphics_common.cpp:
```diff
@@ -0,0 +1,1454 @@
+    return buffer;
+}
+
+#if 0
```
Please no defined-out code either :-)

per-mathisen-arm commented on CMakeLists.txt:
```diff
@@ -63,12 +63,12 @@ endfunction()
+	target_include_directories(vulkan_${ARGV0} PUBLIC ${PROJECT_SOURCE_DIR}/include ${PROJECT_SOURCE_DIR} ${PROJECT_SOURCE_DIR}/external ${PROJECT_SOURCE_DIR}/external/Vulkan-Headers/include)
 	install(TARGETS vulkan_${ARGV0} DESTINATION tests)
 endfunction()
 
```
The vulkan_graphics_1 test is not added to the list of build targets?

per-mathisen-arm commented on src/vulkan_graphics_1.cpp:
```diff
@@ -0,0 +1,399 @@
+#include <chrono>
+
+#define STB_IMAGE_IMPLEMENTATION
+#include <stb_image.h>
```
Should we add this file in external as well, like we do with stb_image_write.h now?

per-mathisen-arm commented on src/vulkan_graphics_common.cpp:
```diff
@@ -0,0 +1,1454 @@
+
+void graphic_usage()
+{
+	printf("-i/--image-output      Save an image of the output to disk\n");
```
Does this work now?

kexia01 replied:
not yet now. removed

kexia01 committed:
Add destroy() function.
Change to use unique pointer as much as possible.
Add asset with include file style.
Fix coding format.

kexia01 committed:
Fix code style to be consistent with codebase.
Add license file in asset/image/ for image usage.
Integrate test case into benchmarking functions.
